cpp/src/arrow/filesystem/hdfs.cc

```diff
@@ -0,0 +1,298 @@
+  }
+
+  Status DeleteFile(const std::string& path) {
+    if (IsDirectory(path)) {
+      return Status::IOError("path is a directory");
+    }
+    RETURN_NOT_OK(client_->Delete(path));
+    return Status::OK();
+  }
+
+  Status Move(const std::string& src, const std::string& dest) {
+    RETURN_NOT_OK(client_->Rename(src, dest));
+    return Status::OK();
+  }
+
+  Status CopyFile(const std::string& src, const std::string& dest) {
+    // TODO
+    return Status::NotImplemented("HadoopFileSystem::CopyFile is not supported yet");
+  }
+
+  Status OpenInputStream(const std::string& path, std::shared_ptr<io::InputStream>* out) {
+    // TODO
+    return Status::NotImplemented(
+        "HadoopFileSystem::OpenInputStream is not supported yet");
+  }
+
+  Status OpenInputFile(const std::string& path,
+                       std::shared_ptr<io::RandomAccessFile>* out) {
+    std::shared_ptr<io::HdfsReadableFile> file;
+    RETURN_NOT_OK(client_->OpenReadable(path, &file));
+    *out = file;
+    return Status::OK();
+  }
+
+  Status OpenOutputStream(const std::string& path,
+                          std::shared_ptr<io::OutputStream>* out) {
+    bool append = false;
+    return OpenOutputStreamGeneric(path, append, out);
+  }
+
+  Status OpenAppendStream(const std::string& path,
+                          std::shared_ptr<io::OutputStream>* out) {
+    bool append = true;
+    return OpenOutputStreamGeneric(path, append, out);
+  }
+
+ protected:
+  HdfsOptions options_;
+  std::shared_ptr<::arrow::io::HadoopFileSystem> client_;
+
+  Status OpenOutputStreamGeneric(const std::string& path, bool append,
+                                 std::shared_ptr<io::OutputStream>* out) {
+    std::shared_ptr<io::HdfsOutputStream> stream;
+    RETURN_NOT_OK(client_->OpenWritable(path, append, options_.buffer_size,
+                                        options_.replication, options_.default_block_size,
+                                        &stream));
+    *out = stream;
+    return Status::OK();
+  }
+
+  bool IsDirectory(const std::string& path) {
+    io::HdfsPathInfo info;
+    Status status = client_->GetPathInfo(path, &info);
+    if (!status.ok()) {
+      return false;
+    }
+    if (info.kind == io::ObjectType::DIRECTORY) {
+      return true;
+    }
+    return false;
+  }
+
+  TimePoint ToTimePoint(int secs) {
+    std::chrono::nanoseconds ns_count(static_cast<int64_t>(secs) * 1000000000);
+    return TimePoint(std::chrono::duration_cast<TimePoint::duration>(ns_count));
+  }
+};
+
+void HdfsOptions::ConfigureEndPoint(const std::string& host, int port) {
+  hdfs_config.host = host;
+  hdfs_config.port = port;
+}
+
+void HdfsOptions::ConfigureHdfsDriver(bool use_hdfs3) {
+  if (use_hdfs3) {
+    hdfs_config.driver = ::arrow::io::HdfsDriver::LIBHDFS3;
+  } else {
+    hdfs_config.driver = ::arrow::io::HdfsDriver::LIBHDFS;
+  }
+}
+
+void HdfsOptions::ConfigureHdfsUser(const std::string& user_name) {
+  if (!user_name.empty()) {
```

Comment:
Why this?

Reply:
My idea is if we don't need to spesify username, then leave it as originally is instead of changing it to "", so if there is some assumption for hdfs_config, for example, if user is null, then use is XXX, I won't make override that.